### ISSUE TITLE
feat(tensor_parallel): add VocabParallelEmbedding

### ISSUE THREAD

**Issue by dylanyunlon** [dylanyunlon/Neuron_SP#118]
## 问题描述
`deepspeed/core/tensor_parallel/layers.py` 中 `VocabParallelEmbedding` 需完成以下 DES-LOC 扩展：
- vocab_size 非整除 TP 时的 dummy-token padding（vocab padding 到 tp 倍数）
- 末 rank 的 padding token embedding 初始化为零并在 all-reduce 前做 mask
- `sharded_state_dict()` 接口对接 dist_checkpointing（weight axis=0 shard）
- 连接 `desloc_engine/autosp` hook

## 技术方案
1. `__init__` 中：`padded_vocab_size = ceil(vocab_size / tp_world_size) * tp_world_size`；各 rank 的 vocab 区间 `[vocab_start, vocab_end)`；末 rank 多出 `padded_vocab_size - vocab_size` 个 dummy token
2. `forward()` 中：对 `input_ids` 做 vocab 区间 mask（超出区间 → 0 embedding，不贡献梯度）；all-reduce 合并各 rank 输出
3. 在 `vocab_start/vocab_end` 属性上加 `@property` 暴露给 transformer 层（用于 logit mask）
4. `sharded_state_dict()`：weight axis=0 打 TP shard
5. 测试：`tests/unit/core/tensor_parallel/test_vocab_parallel_embedding.py`

## 验收条件
```bash
# 1. 导入
python -c "from deepspeed.core.tensor_parallel.layers import VocabParallelEmbedding; print('OK')"

# 2. 无 stub
grep -n "NotImplementedError" deepspeed/core/tensor_parallel/layers.py | grep -i "vocab" && echo "FAIL" || echo "PASS"

# 3. padding 正确性（vocab_size=50257, tp=8 → padded=50264）
python -m pytest tests/unit/core/tensor_parallel/test_vocab_parallel_embedding.py -k "padding" -v

# 4. 数值等价：VocabParallelEmbedding(tp=1) vs nn.Embedding 输出一致
python -m pytest tests/unit/core/tensor_parallel/test_vocab_parallel_embedding.py -k "equivalence" -v

# 5. autosp hook 注册
python -c "from deepspeed.core.hetero_bridge.autosp_hook import register_tp_hooks; print('hook import OK')"
```

## 优先级
P1 — vocab embedding 是 GPT 模型首层，blocking #119 (GPTDataset) 的 end-to-end 测试路径
